apollo-router/src/plugins/authorization/authenticated.rs

```diff
@@ -127,79 +115,91 @@ impl<'a> traverse::Visitor for AuthenticatedCheckVisitor<'a> {
+        executable: &'a ast::Document,
+        implementers_map: &'a HashMap<Name, HashSet<Name>>,
+    ) -> Self {
         Self {
-            compiler,
-            file_id,
+            schema,
+            fragments: transform::collect_fragments(executable),
+            implementers_map,
             query_requires_authentication: false,
             unauthorized_paths: Vec::new(),
             current_path: Path::default(),
         }
     }
 
-    fn is_field_authenticated(&self, field: &FieldDefinition) -> bool {
-        field
-            .directive_by_name(AUTHENTICATED_DIRECTIVE_NAME)
-            .is_some()
-            || field
-                .ty()
-                .type_def(&self.compiler.db)
-                .map(|t| self.is_type_authenticated(&t))
-                .unwrap_or(false)
+    fn is_field_authenticated(&self, field: &schema::FieldDefinition) -> bool {
+        field.directives.has(AUTHENTICATED_DIRECTIVE_NAME)
+            || self
+                .schema
+                .types
+                .get(field.ty.inner_named_type())
+                .is_some_and(|t| self.is_type_authenticated(t))
     }
 
-    fn is_type_authenticated(&self, t: &TypeDefinition) -> bool {
-        t.directive_by_name(AUTHENTICATED_DIRECTIVE_NAME).is_some()
+    fn is_type_authenticated(&self, t: &schema::ExtendedType) -> bool {
+        t.directives().has(AUTHENTICATED_DIRECTIVE_NAME)
     }
 
     fn implementors_with_different_requirements(
         &self,
-        parent_type: &str,
-        node: &hir::Field,
+        field_def: &ast::FieldDefinition,
+        node: &ast::Field,
     ) -> bool {
         // if all selections under the interface field are fragments with type conditions
         // then we don't need to check that they have the same authorization requirements
-        if node.selection_set().fields().is_empty() {
+        if node.selection_set.iter().all(|sel| {
+            matches!(
+                sel,
+                ast::Selection::FragmentSpread(_) | ast::Selection::InlineFragment(_)
+            )
+        }) {
             return false;
         }
 
-        if let Some(type_definition) = get_field_type(self, parent_type, node.name())
-            .and_then(|ty| self.compiler.db.find_type_definition_by_name(ty))
-        {
-            if self.implementors_with_different_type_requirements(&type_definition) {
+        let type_name = field_def.ty.inner_named_type();
+        if let Some(type_definition) = self.schema.types.get(type_name) {
+            if self.implementors_with_different_type_requirements(type_name, type_definition) {
                 return true;
             }
         }
         false
     }
 
-    fn implementors_with_different_type_requirements(&self, t: &TypeDefinition) -> bool {
-        if t.is_interface_type_definition() {
+    fn implementors_with_different_type_requirements(
+        &self,
+        type_name: &str,
+        t: &schema::ExtendedType,
```

Comment:
Yes it’s a bit weird. I’m not sure it should stay that way but right now `ExtendedType` does not contain the type’s name. This is because `Schema` contains an `IndexMap<Name, ExtendedType>`, so the name is already in the map key. With everything being mutable, omitting it in the map value eliminates the possibility of inconsistency.

However this pattern is not applied everywhere. For example object fields and type definitions are in a name-indexed map but include a name in the map value. We should pick one pattern or the other and stick to it.

Reply:
ok, I see